File: macos/Sources/ElevateCore/Catalogue/ManualRoleSource.swift
import Foundation

/// A role the user asserts they hold in a tenant where discovery is unavailable.
public struct ManualRole: Codable, Hashable, Sendable {
    public var tenantKey: TenantKey
    public var scope: RoleScope
    public var displayName: String
    public init(tenantKey: TenantKey, scope: RoleScope, displayName: String) {
        self.tenantKey = tenantKey
        self.scope = scope
        self.displayName = displayName
    }
}

public enum ManualRoleSource {
    public static func eligibleRoles(from manual: [ManualRole], tenantKey: TenantKey) -> [EligibleRole] {
        manual.filter { $0.tenantKey == tenantKey }.map {
            let detail: String? = switch $0.scope {
            case .azureResource(let scope, _): scope
            case .group(_, let accessId): accessId == .owner ? "owner" : "member"
            default: nil
            }
            return EligibleRole(key: RoleKey(identityId: tenantKey.identityId, tenantId: tenantKey.tenantId, scope: $0.scope),
                                displayName: $0.displayName, detail: detail, source: .manual, policy: .manualDefault)
        }
    }

    /// Discovered roles win over manual entries with the same key; a manual Azure entry is also dropped when a
    /// discovered Azure role has the same scope and display name (the manual entry names the role, ARM ids it).
    public static func merge(discovered: [EligibleRole], manual: [EligibleRole]) -> [EligibleRole] {
        let known = Set(discovered.map(\.key))
        let azureNames = Set(discovered.compactMap { role -> String? in
            guard case .azureResource(let scope, _) = role.key.scope else { return nil }
            return scope.lowercased() + "|" + role.displayName.lowercased()
        })
        return discovered + manual.filter { role in
            guard !known.contains(role.key) else { return false }
            if case .azureResource(let scope, _) = role.key.scope {
                return !azureNames.contains(scope.lowercased() + "|" + role.displayName.lowercased())
            }
            return true
        }
    }
}

File: windows/src/Elevate.Core/Catalogue/ManualRoleSource.cs
using Elevate.Core.Models;

namespace Elevate.Core.Catalogue;

/// <summary>A role the user asserts they hold in a tenant where discovery is unavailable.</summary>
public sealed record ManualRole(TenantKey TenantKey, RoleScope Scope, string DisplayName);

public static class ManualRoleSource
{
    /// <summary>The manual roles of one tenant, as eligible roles carrying the default policy.</summary>
    public static IReadOnlyList<EligibleRole> EligibleRoles(IEnumerable<ManualRole> manual, TenantKey tenantKey)
    {
        ArgumentNullException.ThrowIfNull(manual);

        return
        [
            .. manual.Where(r => r.TenantKey == tenantKey).Select(r => new EligibleRole(
                new RoleKey(tenantKey.IdentityId, tenantKey.TenantId, r.Scope),
                r.DisplayName,
                RoleSource.Manual,
                RolePolicy.ManualDefault,
                Detail(r.Scope))),
        ];
    }

    private static string? Detail(RoleScope scope) => scope switch
    {
        AzureResourceScope azure => azure.Scope,
        GroupScope group => group.AccessId == GroupAccess.Owner ? "owner" : "member",
        _ => null,
    };

    /// <summary>
    /// Discovered roles win over manual entries with the same key; a manual Azure entry is also dropped when a
    /// discovered Azure role has the same scope and display name (the manual entry names the role, ARM ids it).
    /// </summary>
    public static IReadOnlyList<EligibleRole> Merge(IEnumerable<EligibleRole> discovered, IEnumerable<EligibleRole> manual)
    {
        ArgumentNullException.ThrowIfNull(discovered);
        ArgumentNullException.ThrowIfNull(manual);

        var discoveredRoles = discovered.ToList();
        var known = new HashSet<RoleKey>(discoveredRoles.Select(r => r.Key));
        var azureNames = new HashSet<string>(discoveredRoles
            .Where(r => r.Key.Scope is AzureResourceScope)
            .Select(r => AzureName(((AzureResourceScope)r.Key.Scope).Scope, r.DisplayName)));

        return
        [
            .. discoveredRoles,
            .. manual.Where(role => !known.Contains(role.Key)
                && (role.Key.Scope is not AzureResourceScope azure
                    || !azureNames.Contains(AzureName(azure.Scope, role.DisplayName)))),
        ];
    }

    private static string AzureName(string scope, string displayName) =>
        scope.ToLowerInvariant() + "|" + displayName.ToLowerInvariant();
}
